C ffi docs update

lygstate opened the pull request:
<!-- homu-ignore:start -->
<!--
If this PR is related to an unstable feature or an otherwise tracked effort,
please link to the relevant tracking issue here. If you don't know of a related
tracking issue or there are none, feel free to ignore this.

This PR will get automatically assigned to a reviewer. In case you would like
a specific user to review your work, you can assign it to them by using

    r? <reviewer name>
-->
<!-- homu-ignore:end -->
No functional change, just document update

lygstate committed:
c ffi document fixes for c_short.md and c_longlong.md

rust-log-analyzer commented:
The job **`aarch64-gnu-llvm-20-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/17897972864/job/50887117097) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/50887117097) [(plain)](https://github.com/rust-lang/rust/commit/8de8b4fcb0ed50725afaaa963209c591cd9ce4fe/checks/50887117097/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
  --> library/core/src/ffi/primitives.rs:14:9
   |
 6 | / macro_rules! type_alias {
 7 | |     {
 8 | |       $Docfile:tt, $Alias:ident = $Real:ty;
 9 | |       $( $Cfg:tt )*
...  |
14 | |         pub type $Alias = $Real;
   | |         ^^^^^^^^^^^^^^^^^^^^^^^^
15 | |     }
16 | | }
   | |_- in this expansion of `type_alias!`
17 |
18 |   type_alias! { "c_char.md", c_char = c_char_definition::c_char; #[doc(cfg(all()))] }
   |   ----------------------------------------------------------------------------------- in this macro invocation

error: type alias has missing stability attribute
  --> library/core/src/ffi/primitives.rs:14:9
   |
 6 | / macro_rules! type_alias {
 7 | |     {
 8 | |       $Docfile:tt, $Alias:ident = $Real:ty;
 9 | |       $( $Cfg:tt )*
...  |
14 | |         pub type $Alias = $Real;
   | |         ^^^^^^^^^^^^^^^^^^^^^^^^
15 | |     }
16 | | }
   | |_- in this expansion of `type_alias!`
...
20 |   type_alias! { "c_schar.md", c_schar = i8; }
   |   ------------------------------------------- in this macro invocation

error: type alias has missing stability attribute
  --> library/core/src/ffi/primitives.rs:14:9
   |
 6 | / macro_rules! type_alias {
 7 | |     {
 8 | |       $Docfile:tt, $Alias:ident = $Real:ty;
 9 | |       $( $Cfg:tt )*
...  |
14 | |         pub type $Alias = $Real;
   | |         ^^^^^^^^^^^^^^^^^^^^^^^^
15 | |     }
16 | | }
   | |_- in this expansion of `type_alias!`
...
21 |   type_alias! { "c_uchar.md", c_uchar = u8; }
   |   ------------------------------------------- in this macro invocation

error: type alias has missing stability attribute
  --> library/core/src/ffi/primitives.rs:14:9
   |
 6 | / macro_rules! type_alias {
 7 | |     {
 8 | |       $Docfile:tt, $Alias:ident = $Real:ty;
 9 | |       $( $Cfg:tt )*
...  |
14 | |         pub type $Alias = $Real;
   | |         ^^^^^^^^^^^^^^^^^^^^^^^^
15 | |     }
16 | | }
   | |_- in this expansion of `type_alias!`
...
22 |   type_alias! { "c_short.md", c_short = i16; }
   |   -------------------------------------------- in this macro invocation

error: type alias has missing stability attribute
  --> library/core/src/ffi/primitives.rs:14:9
   |
 6 | / macro_rules! type_alias {
 7 | |     {
 8 | |       $Docfile:tt, $Alias:ident = $Real:ty;
 9 | |       $( $Cfg:tt )*
...  |
14 | |         pub type $Alias = $Real;
   | |         ^^^^^^^^^^^^^^^^^^^^^^^^
15 | |     }
16 | | }
   | |_- in this expansion of `type_alias!`
...
23 |   type_alias! { "c_ushort.md", c_ushort = u16; }
   |   ---------------------------------------------- in this macro invocation

error: type alias has missing stability attribute
  --> library/core/src/ffi/primitives.rs:14:9
   |
 6 | / macro_rules! type_alias {
 7 | |     {
 8 | |       $Docfile:tt, $Alias:ident = $Real:ty;
 9 | |       $( $Cfg:tt )*
...  |
14 | |         pub type $Alias = $Real;
   | |         ^^^^^^^^^^^^^^^^^^^^^^^^
15 | |     }
16 | | }
   | |_- in this expansion of `type_alias!`
...
25 |   type_alias! { "c_int.md", c_int = c_int_definition::c_int; #[doc(cfg(all()))] }
   |   ------------------------------------------------------------------------------- in this macro invocation

error: type alias has missing stability attribute
  --> library/core/src/ffi/primitives.rs:14:9
   |
 6 | / macro_rules! type_alias {
 7 | |     {
 8 | |       $Docfile:tt, $Alias:ident = $Real:ty;
 9 | |       $( $Cfg:tt )*
...  |
14 | |         pub type $Alias = $Real;
   | |         ^^^^^^^^^^^^^^^^^^^^^^^^
15 | |     }
16 | | }
   | |_- in this expansion of `type_alias!`
...
26 |   type_alias! { "c_uint.md", c_uint = c_int_definition::c_uint; #[doc(cfg(all()))] }
   |   ---------------------------------------------------------------------------------- in this macro invocation

error: type alias has missing stability attribute
  --> library/core/src/ffi/primitives.rs:14:9
   |
 6 | / macro_rules! type_alias {
 7 | |     {
 8 | |       $Docfile:tt, $Alias:ident = $Real:ty;
 9 | |       $( $Cfg:tt )*
...  |
14 | |         pub type $Alias = $Real;
   | |         ^^^^^^^^^^^^^^^^^^^^^^^^
15 | |     }
16 | | }
   | |_- in this expansion of `type_alias!`
...
28 |   type_alias! { "c_long.md", c_long = c_long_definition::c_long; #[doc(cfg(all()))] }
   |   ----------------------------------------------------------------------------------- in this macro invocation

error: type alias has missing stability attribute
  --> library/core/src/ffi/primitives.rs:14:9
   |
 6 | / macro_rules! type_alias {
 7 | |     {
 8 | |       $Docfile:tt, $Alias:ident = $Real:ty;
 9 | |       $( $Cfg:tt )*
...  |
14 | |         pub type $Alias = $Real;
   | |         ^^^^^^^^^^^^^^^^^^^^^^^^
15 | |     }
16 | | }
   | |_- in this expansion of `type_alias!`
...
29 |   type_alias! { "c_ulong.md", c_ulong = c_long_definition::c_ulong; #[doc(cfg(all()))] }
   |   -------------------------------------------------------------------------------------- in this macro invocation

error: type alias has missing stability attribute
  --> library/core/src/ffi/primitives.rs:14:9
   |
 6 | / macro_rules! type_alias {
 7 | |     {
 8 | |       $Docfile:tt, $Alias:ident = $Real:ty;
 9 | |       $( $Cfg:tt )*
...  |
14 | |         pub type $Alias = $Real;
   | |         ^^^^^^^^^^^^^^^^^^^^^^^^
15 | |     }
16 | | }
   | |_- in this expansion of `type_alias!`
...
31 |   type_alias! { "c_longlong.md", c_longlong = i64; }
   |   -------------------------------------------------- in this macro invocation

error: type alias has missing stability attribute
  --> library/core/src/ffi/primitives.rs:14:9
   |
 6 | / macro_rules! type_alias {
 7 | |     {
 8 | |       $Docfile:tt, $Alias:ident = $Real:ty;
 9 | |       $( $Cfg:tt )*
...  |
14 | |         pub type $Alias = $Real;
   | |         ^^^^^^^^^^^^^^^^^^^^^^^^
15 | |     }
16 | | }
   | |_- in this expansion of `type_alias!`
...
32 |   type_alias! { "c_ulonglong.md", c_ulonglong = u64; }
   |   ---------------------------------------------------- in this macro invocation

error: type alias has missing stability attribute
  --> library/core/src/ffi/primitives.rs:14:9
   |
 6 | / macro_rules! type_alias {
 7 | |     {
 8 | |       $Docfile:tt, $Alias:ident = $Real:ty;
 9 | |       $( $Cfg:tt )*
...  |
14 | |         pub type $Alias = $Real;
   | |         ^^^^^^^^^^^^^^^^^^^^^^^^
15 | |     }
16 | | }
   | |_- in this expansion of `type_alias!`
...
34 |   type_alias! { "c_float.md", c_float = f32; }
   |   -------------------------------------------- in this macro invocation

error: type alias has missing stability attribute
  --> library/core/src/ffi/primitives.rs:14:9
   |
 6 | / macro_rules! type_alias {
 7 | |     {
 8 | |       $Docfile:tt, $Alias:ident = $Real:ty;
 9 | |       $( $Cfg:tt )*
...  |
14 | |         pub type $Alias = $Real;
   | |         ^^^^^^^^^^^^^^^^^^^^^^^^
15 | |     }
16 | | }
   | |_- in this expansion of `type_alias!`
...
35 |   type_alias! { "c_double.md", c_double = f64; }
   |   ---------------------------------------------- in this macro invocation

error: type alias has missing stability attribute
   --> library/core/src/ffi/primitives.rs:14:9
    |
  6 | / macro_rules! type_alias {
  7 | |     {
  8 | |       $Docfile:tt, $Alias:ident = $Real:ty;
  9 | |       $( $Cfg:tt )*
...   |
 14 | |         pub type $Alias = $Real;
    | |         ^^^^^^^^^^^^^^^^^^^^^^^^
 15 | |     }
 16 | | }
    | |_- in this expansion of `type_alias!`
...
153 |   type_alias! { "c_size_t.md", c_size_t = usize; }
    |   ------------------------------------------------ in this macro invocation

error: type alias has missing stability attribute
   --> library/core/src/ffi/primitives.rs:14:9
    |
  6 | / macro_rules! type_alias {
  7 | |     {
  8 | |       $Docfile:tt, $Alias:ident = $Real:ty;
  9 | |       $( $Cfg:tt )*
...   |
 14 | |         pub type $Alias = $Real;
    | |         ^^^^^^^^^^^^^^^^^^^^^^^^
 15 | |     }
 16 | | }
    | |_- in this expansion of `type_alias!`
...
154 |   type_alias! { "c_ptrdiff_t.md", c_ptrdiff_t = isize; }
    |   ------------------------------------------------------ in this macro invocation

error: type alias has missing stability attribute
   --> library/core/src/ffi/primitives.rs:14:9
    |
  6 | / macro_rules! type_alias {
  7 | |     {
  8 | |       $Docfile:tt, $Alias:ident = $Real:ty;
  9 | |       $( $Cfg:tt )*
...   |
 14 | |         pub type $Alias = $Real;
    | |         ^^^^^^^^^^^^^^^^^^^^^^^^
 15 | |     }
 16 | | }
    | |_- in this expansion of `type_alias!`
...
155 |   type_alias! { "c_ssize_t.md", c_ssize_t = isize; }
    |   -------------------------------------------------- in this macro invocation

[RUSTC-TIMING] core test:false 25.981
error: could not compile `core` (lib) due to 16 previous errors
Bootstrap failed while executing `--stage 2 test --skip compiler --skip src`
```

</details>

lygstate commented on library/core/src/ffi/primitives.rs:
```diff
@@ -150,26 +150,9 @@ mod c_long_definition {
-/// platforms where this is not the case.
-#[unstable(feature = "c_size_t", issue = "88345")]
-pub type c_ssize_t = isize;
+type_alias! { "c_size_t.md", c_size_t = usize; }
+type_alias! { "c_ptrdiff_t.md", c_ptrdiff_t = isize; }
+type_alias! { "c_ssize_t.md", c_ssize_t = isize; }
```
Yeap, this is the question I raised at https://github.com/rust-lang/rust/pull/136334#discussion_r2366638706

lygstate committed:
Avoid duplicate stable and unstable attributes between ffi/mod.rs and ffi/primitives.rs

Split c_ptrdiff_t.md c_size_t.md c_ssize_t.md out of primitives.rs for consistence

lygstate commented:
@ricci009 take a look at this, I think this may simplify the MR https://github.com/rust-lang/rust/pull/133944

Mark-Simulacrum commented on library/core/src/ffi/mod.rs:
```diff
@@ -38,13 +38,42 @@ pub use self::va_list::{VaArgSafe, VaList, VaListImpl};
+        #[doc = include_str!($Docfile)]
+        #[doc(cfg(all()))]
+        $( $Cfg )*
+        pub type $Alias = $Real;
```
If you'd like to revise this PR to not move things around and keep the old structure, I'd be happy to merge it with just the new docs. (I think that means just keeping the first commit). The separate files are intentional per https://github.com/rust-lang/rust/pull/136334, though we haven't (yet) made use of that, so I'd prefer to avoid significant refactoring in the shape for now.

Dylan-DPC commented:
Closing this as inactive. Feel free to reöpen this pr or create a new pr if you get the time to work on this. Thanks
